src/pool/poolClient.ts

```diff
@@ -362,6 +414,46 @@ export class Client {
   private getOrCreatePoolEvents() {
     return this.poolEvents
   }
+  private getOrCreateUserService(userAddress:string,tokenAddress:string){
+    if(has(this.userServices,[tokenAddress,userAddress])) return get(this.userServices,[tokenAddress,userAddress]);
+    const erc20Contract = this.getOrCreateErc20Contract(tokenAddress)
+    const userService = new UserState(erc20Contract);
+    // this service is stateful now, so needs to be cached
+    set(this.userServices,[tokenAddress,userAddress],userService)
+    return userService
+  }
+  private updateExchangeRateTable(l1TokenAddress:string, exchangeRateTable:Record<number,BigNumberish>):Record<number,BigNumberish>{
+    if(!this.exchangeRateTable[l1TokenAddress]) this.exchangeRateTable[l1TokenAddress] = {}
+    this.exchangeRateTable[l1TokenAddress] = {...this.exchangeRateTable[l1TokenAddress], ...exchangeRateTable}
+    return this.exchangeRateTable[l1TokenAddress]
+  }
+  // calculates the value of each LP token transfer at the block it was sent. this only works if we have archive node
+  async calculateLpTransferValue(l1TokenAddress:string, userState: Awaited<ReturnType<UserState["read"]>>){
+    assert(this.config.hasArchive,'Can only calculate historical lp values with archive node')
+    const contract = this.getOrCreatePoolContract();
+    const pool = new PoolState(contract,this.config.hubPoolAddress);
+    const blockNumbers = userState.transferEvents
+      .map(x=>x.blockNumber)
+      .filter(blockNumber=>!this.exchangeRateTable?.[l1TokenAddress]?.[blockNumber]);
```

Comment:
Shouldn't the exchange rate table and the object at the l1Token address always exist? I thought that's what the code above was meant to ensure. If that's the case, then why do we need all the optional chaining here?